### PR TITLE
fix: fix flay migrate_database_files_from_one_server_to_another and prepare for reuse + fix

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/3311

This fixes a flaky test and prepares much of it for reuse in https://github.com/influxdata/influxdb_iox/pull/3292

I will comment inline about the actual fix

### CODE REVIEW

File: influxdb_iox/tests/end_to_end_cases/database_migration.rs [hunk 1]
```diff
@@ -1,18 +1,47 @@
...
+    let target_dir = target_dir.as_ref();
+
+    // if the target database has had no data written yet, it will
+    // have no `data` directory
+    println!("Ensuring {:?} directory exists", target_dir);
+    Command::new("mkdir")
+        .arg("-p")
+        .arg(target_dir.to_string_lossy().to_string())
+        .assert()
+        .success();
```
Comment (alamb) [inline influxdb_iox/tests/end_to_end_cases/database_migration.rs:28]: This is the fix for https://github.com/influxdata/influxdb_iox/pull/3292 (if the target directory doesn't exist the files end up in the wrong spot)

File: influxdb_iox/tests/end_to_end_cases/database_migration.rs [hunk 2]
```diff
@@ -1,18 +1,47 @@
...
-async fn migrate_database_files_from_one_server_to_another() {
+/// Copy the `source_dir` directory into the `target_dir` directory using the `cp` command
+fn cp_dir(source_dir: impl AsRef<Path>, target_dir: impl AsRef<Path>) {
+    let source_dir = source_dir.as_ref();
+    let target_dir = target_dir.as_ref();
+
+    // needed so that if the target server has had no databases
+    // created yet, it will have no `data` directory. See #3292
+    println!("Ensuring {:?} directory exists", target_dir);
+    Command::new("mkdir")
```
Comment (alamb) [inline influxdb_iox/tests/end_to_end_cases/database_migration.rs:24]: This is the fix for https://github.com/influxdata/influxdb_iox/pull/3292 (if the target directory doesn't exist the files end up in the wrong spot)

The rest of this PR is just refactoring into reusable pieces I am using for a new test

**Review by e-dard — APPROVED** [review]

File: influxdb_iox/tests/end_to_end_cases/database_migration.rs (hunk 2 above)
Comment (crepererum) [reply influxdb_iox/tests/end_to_end_cases/database_migration.rs:24]: Do we really need to shell out to `mkdir` and `cp` to create an copy a bit of FS content? Should the [stdlib](https://doc.rust-lang.org/std/fs/index.html#functions) be sufficient for that?

Comment (alamb) [reply influxdb_iox/tests/end_to_end_cases/database_migration.rs:24]: I couldn't find anything in the stdlib that will do `cp -R` (aka recursive directory copy) and I thought shelling out to `cp` was less icky than importing a new crate.

**Pull request merged**